### PR TITLE
Add two more neon functions

### PR DESCRIPTION
**Pull request by gizmo98**
-gSPTransformVertex_NEON. Neon intrinsic port of https://github.com/ricrpi/mupen64plus-video-gles2n64/blob/master/src/gSPNeon.cpp#L409.
-RSP_LoadMatrix neon intrinsic version.
-Move x86 asm version of RSP_LoadMatrix into an own file.
-Add files to project files.

@gonetz 
I only tested cmake.

### CODE REVIEW

**Comment by fzurita** [comment]
The new inline neon assembly compiles just fine in Android. The X86 inline assembly is giving me build errors though. I probably should figure out what flags to use to enable it.

**Comment by gizmo98** [comment]
I can disable it for now. X86_ASM and WIN32_ASM defines were not used before.

File: src/gSPNeon.cpp [hunk 1]
```diff
@@ -113,3 +114,23 @@ void gSPBillboardVertex4NEON(u32 v)
...
+    float32x4_t _mtx3 = vld1q_f32(mtx[3]);
+
+    // Multiply and add
+    _mtx0 = vmlaq_n_f32(_mtx3, _mtx0, _vtx[0]);    // _mtx0 = _mtx3 + _mtx0 * _vtx[0]
+    _mtx0 = vmlaq_n_f32(_mtx0, _mtx1, _vtx[1]);    // _mtx0 = _mtx0 + _mtx1 * _vtx[1]
+    _mtx0 = vmlaq_n_f32(_mtx0, _mtx2, _vtx[2]);    // _mtx0 = _mtx0 + _mtx2 * _vtx[2]
+
+    // Store vtx
+    vst1q_f32(vtx, _mtx0);
+}
```
Comment (gonetz) [inline src/gSPNeon.cpp]: * No newline at end of file

**Comment by gonetz** [comment]
There are several *Neon.cpp files. These files not necessary for desktop build. Lets move them all to special folder, say 'ARM_Neon' or just 'Neon'.

**Comment by gonetz** [comment]
Unless you not planning to implement more gSP functions, gSPNeon.cpp better to rename to TransformVertexNeon.cpp

File: src/RSP_LoadMatrix.cpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        SHORT integer[4][4];
+        WORD fraction[4][4];
+    } *n64Mat = (struct _N64Matrix *)&RDRAM[address];
+
+    int i, j;
+
+    for (i = 0; i < 4; i++)
+        for (j = 0; j < 4; j++)
+            mtx[i][j] = (f32)(n64Mat->integer[i][j^1]) + (f32)(n64Mat->fraction[i][j^1]) * recip;
+}
```
Comment (gonetz) [inline src/RSP_LoadMatrix.cpp]: No newline at end of file

File: src/RSP_LoadMatrixX86.cpp [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+    "    add     esi, 0x08"                      "\n\t"
+    "    add     edi, 0x10"                      "\n\t"
+    "    loop    LoadLoop"                       "\n\t"
+    ".att_syntax prefix"                         "\n\t"
+    : /* no output */
+    : "f"(recip), "S"((int)RDRAM+address), "D"(mtx), "c"(4)
+    : "memory" );
+#endif // WIN32_ASM
+}
```
Comment (gonetz) [inline src/RSP_LoadMatrixX86.cpp]: No newline at end of file

File: src/RSP_LoadMatrix.cpp [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+#include "Combiner.h"
+#include "FrameBuffer.h"
+#include "DepthBuffer.h"
+#include "FrameBufferInfo.h"
+#include "GBI.h"
+#include "PluginAPI.h"
+#include "Config.h"
+#include "TextureFilterHandler.h"
+#include "DisplayWindow.h"
+
```
Comment (gonetz) [inline src/RSP_LoadMatrix.cpp]: All these includes look redundant.

File: src/RSP_LoadMatrix.cpp [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+using namespace std;
+
```
Comment (gonetz) [inline src/RSP_LoadMatrix.cpp]: std is not used below

File: src/RSP_LoadMatrixNeon.cpp [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+#include "FrameBuffer.h"
+#include "DepthBuffer.h"
+#include "FrameBufferInfo.h"
+#include "GBI.h"
+#include "PluginAPI.h"
+#include "Config.h"
+#include "TextureFilterHandler.h"
+#include "DisplayWindow.h"
+#include "arm_neon.h"
+
```
Comment (gonetz) [inline src/RSP_LoadMatrixNeon.cpp]: All these includes look redundant.

File: src/RSP_LoadMatrixNeon.cpp [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+using namespace std;
+
```
Comment (gonetz) [inline src/RSP_LoadMatrixNeon.cpp]: std not used below

File: src/RSP_LoadMatrixX86.cpp [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+#include "Combiner.h"
+#include "FrameBuffer.h"
+#include "DepthBuffer.h"
+#include "FrameBufferInfo.h"
+#include "GBI.h"
+#include "PluginAPI.h"
+#include "Config.h"
+#include "TextureFilterHandler.h"
+#include "DisplayWindow.h"
+
```
Comment (gonetz) [inline src/RSP_LoadMatrixX86.cpp]: All these includes look redundant.

File: src/RSP_LoadMatrixX86.cpp [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+using namespace std;
+
```
Comment (gonetz) [inline src/RSP_LoadMatrixX86.cpp]: std not used below

File: src/mupen64plus-video-gliden64.mk [hunk 10]
```diff
@@ -126,15 +126,17 @@ MY_LOCAL_LDLIBS := -llog -latomic -lEGL
...
+    MY_LOCAL_SRC_FILES += $(SRCDIR)/gSPNeon.cpp
+    MY_LOCAL_SRC_FILES += $(SRCDIR)/RSP_LoadMatrixNeon.cpp
     MY_LOCAL_CFLAGS += -D__NEON_OPT
     MY_LOCAL_CFLAGS += -D__VEC4_OPT -mfpu=neon -mfloat-abi=softfp -ftree-vectorize -funsafe-math-optimizations -fno-finite-math-only
 
 else ifeq ($(TARGET_ARCH_ABI), x86)
 #    MY_LOCAL_CFLAGS += -DX86_ASM
     MY_LOCAL_CFLAGS += -D__VEC4_OPT
     MY_LOCAL_SRC_FILES += $(SRCDIR)/3DMath.cpp
+    MY_LOCAL_SRC_FILES += $(SRCDIR)/RSP_LoadMatrixX86.cpp
```
Comment (gonetz) [inline src/mupen64plus-video-gliden64.mk]: Let fzurita add it later.

File: src/RSP_LoadMatrixNeon.cpp (hunk 6 above)
Comment (gizmo98) [inline src/RSP_LoadMatrixNeon.cpp]: Moving most includes into gSP.h should be ok?

File: src/RSP_LoadMatrix.cpp (hunk 4 above)
Comment (fzurita) [inline src/RSP_LoadMatrix.cpp]: Android Studio will grey out unnecessary includes. It would help with things such is this.

**Comment by fzurita** [comment]
Also, if we are moving files around, maybe we should consider moving all the ucode implementations into a separate folder.

File: src/RSP_LoadMatrixNeon.cpp (hunk 6 above)
Comment (gonetz) [inline src/RSP_LoadMatrixNeon.cpp]: Why? These includes just useless there. Except for #include "arm_neon.h" in this case.
gSP.h don't need these includes too.

**Comment by gonetz** [comment]
> maybe we should consider moving all the ucode implementations into a separate folder.

Maybe. Would not hurt anything.

**Comment by fzurita** [comment]
I did some more research as to why x86 assembly is not compiling with clang. It's because clang is not behaving exactly the same as gcc for inline assembly when using ```.intel_syntax noprefix```. Some of the assembly would have to be re-written.

**Comment by gonetz** [comment]
> It's probably easier to just not include x86 assembly for Android for now.

If Android device is loaded with x86 CPU, it should be fast enough to run without asm help. I'm not sure does it still have any sense on desktop.

**Comment by gizmo98** [comment]
Should be ok now.

**Comment by fzurita** [comment]
Everything builds ok in Android.

**Comment by fzurita** [comment]
It still runs ok as well.

**Comment by gonetz** [comment]
I fixed few things and merged: 8f49d6b3a4a

**Pull request closed without merging**